### PR TITLE
Allow configuring DKIM options on Vertx's Mail client

### LINKED ISSUE #25247: Mailer - Support DKIM
### Description

As you may know, DKIM is an email authentication method to detect email spoofing. So it's a big deal for those who are security wise.

### Implementation ideas

Currently, the Mailer extension uses Vertx's Mail client which supports DKIM signing. It's just creating a configuration that allows you to configure the DKIM options and pass it to this client.

### PR DESCRIPTION
**Pull request by ejba**
Closes #25247

### CODE REVIEW

**Comment by ejba** [comment]
It's just a draft, so I can get feedback from maintainers. :) 

I still have to figure how to make tests for this.

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/DkimSignOptionsConfig.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    @ConfigItem
+    public Optional<CanonicalizationAlgorithmOption> bodyCanonAlgo;
+
+    /**
+     * Configures the body limit to sign.
+     *
+     * Must be greater than zero.
+     */
+    @ConfigItem
+    public Optional<Integer> bodyLimit;
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/DkimSignOptionsConfig.java:65]: Use `OptionalInt` instead

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/DkimSignOptionsConfig.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    @ConfigItem
+    public Optional<Long> expireTime;
+
+    /**
+     * Configures the signed headers in DKIM, separated by commas.
+     *
+     * The order in the list matters.
+     */
+    @ConfigItem
+    public Optional<String> signedHeaders;
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/DkimSignOptionsConfig.java:87]: This can be an `Optional<List<String>>` instead

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 3]
```diff
@@ -56,10 +60,90 @@ public void stop() {
...
+            }
+            vertxDkimOptions.setSignAlgo(selectedAlgo);
+        }
+
+        if (optionsConfig.signatureTimestamp.isPresent()) {
+            vertxDkimOptions.setSignatureTimestamp(optionsConfig.signatureTimestamp.get());
+        }
+
+        if (optionsConfig.signedHeaders.isPresent()) {
+            List<String> headers = optionsConfig.signedHeaders.get()
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:131]: This shouldn't be needed anymore

Comment (ejba) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:131]: `ConfigItem` deals with blank strings? I didn't know.

**Comment by ejba** [comment]
I noticed the unit tests and integration tests are disabled in `integration-tests/mailer` module.
Can I activate them? Or should I execute them in a different way?

**Comment by gastaldi** [comment]
> I noticed the unit tests and integration tests are disabled in `integration-tests/mailer` module. Can I activate them? Or should I execute them in a different way?

That's odd, are they passing when executing them locally? I am getting: 

```
2022-04-29 09:56:19,023 WARN  [org.tes.con.wai.str.HttpWaitStrategy] (pool-3-thread-1) Unexpected error occurred - will proceed to try to wait anyway: java.lang.IllegalStateException: Target port 9155 is not exposed
	at org.testcontainers.containers.wait.strategy.HttpWaitStrategy.lambda$waitUntilReady$2(HttpWaitStrategy.java:197)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.testcontainers.containers.wait.strategy.HttpWaitStrategy.waitUntilReady(HttpWaitStrategy.java:197)
	at org.testcontainers.containers.wait.strategy.AbstractWaitStrategy.waitUntilReady(AbstractWaitStrategy.java:51)
	at org.testcontainers.containers.GenericContainer.waitUntilContainerStarted(GenericContainer.java:926)
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:480)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:340)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:338)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:326)
	at io.quarkus.it.mailer.FakeMailerTestResource.start(FakeMailerTestResource.java:22)
	at io.quarkus.test.common.TestResourceManager$TestResourceEntryRunnable.run(TestResourceManager.java:452)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

```

**Comment by ejba** [comment]
I am getting the same warning as well. 

However all tests are passing when executing them locally. 
```
INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 26.525 s - in io.quarkus.it.mailer.MailerTest
2022-04-29 13:59:02,556 ERROR [io.ver.ext.mai.imp.SMTPConnection] (vert.x-eventloop-thread-1) socket was closed unexpected.
2022-04-29 13:59:02,558 ERROR [io.ver.ext.mai.imp.SMTPConnection] (vert.x-eventloop-thread-0) socket was closed unexpected.
2022-04-29 13:59:02,604 INFO  [io.quarkus] (main) quarkus-integration-test-mailer stopped in 0.064s
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:build (default) @ quarkus-integration-test-mailer ---
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 2369ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  46.939 s
[INFO] Finished at: 2022-04-29T13:59:06+01:00
```

**Comment by gastaldi** [comment]
I'd say go ahead and enable them, unless @cescoffier has any objections

**Comment by ejba** [comment]
Oh, I noticed now that there is a profile to enable them back. 
`mvn clean install  -f integration-tests/mailer -P test-mailer` does the trick.

That's odd, I didn't find other modules with similar config.

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 4]
```diff
@@ -56,10 +59,94 @@ public void stop() {
...
+
+        if (optionsConfig.signatureTimestamp.isPresent()) {
+            vertxDkimOptions.setSignatureTimestamp(optionsConfig.signatureTimestamp.get());
+        }
+
+        if (optionsConfig.signedHeaders.isPresent()) {
+            List<String> headers = optionsConfig.signedHeaders.get();
+
+            if (headers.size() == 0) {
+                throw new ConfigurationException("Signed headers must have more than zero elements.");
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:133]:
```suggestion
                throw new ConfigurationException("Signed headers must have at least one element.");
```

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 5]
```diff
@@ -56,10 +59,94 @@ public void stop() {
...
+            }
+
+            headers.stream()
+                    .filter(header -> header.equalsIgnoreCase("from"))
+                    .findAny()
+                    .orElseThrow(() -> {
+                        throw new ConfigurationException(
+                                "The \"From\" header must be always be included to the list of headers to sign.");
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:141]:
```suggestion
                                "The \"From\" header must be always included to the list of headers to sign.");
```

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 6]
```diff
@@ -56,10 +59,90 @@ public void stop() {
...
+        if (optionsConfig.signedHeaders.isPresent()) {
+            List<String> headers = optionsConfig.signedHeaders.get();
+
+            headers.stream()
+                    .filter(header -> header.equalsIgnoreCase("from"))
+                    .findAny()
+                    .orElseThrow(() -> {
+                        throw new ConfigurationException(
+                                "The \"From\" header must always be included to the list of headers to sign.");
+                    });
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:138]: You can simplify to: 

```suggestion
            if (headers.stream().noneMatch(header -> header.equalsIgnoreCase("from"))) {
                        throw new ConfigurationException(
                                "The \"From\" header must always be included to the list of headers to sign.");
            }

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 7]
```diff
@@ -56,10 +59,79 @@ public void stop() {
...
+        if (optionsConfig.expireTime.isPresent()) {
+            long expireTime = optionsConfig.expireTime.getAsLong();
+            if (expireTime <= 0) {
+                throw new ConfigurationException("Expirate time must be greater than zero.");
+            }
+            vertxDkimOptions.setExpireTime(expireTime);
+        }
+
+        if (optionsConfig.headerCanonAlgo.isPresent()) {
+            vertxDkimOptions.setBodyCanonAlgo(CanonicalizationAlgorithm.valueOf(optionsConfig.bodyCanonAlgo.get().name()));
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:108]:
```suggestion
            vertxDkimOptions.setHeaderCanonAlgo(CanonicalizationAlgorithm.valueOf(optionsConfig.headerCanonAlgo.get().name()));
```

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 8]
```diff
@@ -56,10 +59,71 @@ public void stop() {
...
+            if (bodyLimit <= 0) {
+                throw new ConfigurationException("Body limit must be greater than zero.");
+            }
+            vertxDkimOptions.setBodyLimit(bodyLimit);
+        }
+
+        if (optionsConfig.expireTime.isPresent()) {
+            long expireTime = optionsConfig.expireTime.getAsLong();
+            if (expireTime <= 0) {
+                throw new ConfigurationException("Expirate time must be greater than zero.");
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:90]:
```suggestion
                throw new ConfigurationException("Expire time must be greater than zero.");
```

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 9]
```diff
@@ -56,10 +59,71 @@ public void stop() {
...
+        if (optionsConfig.auid.isPresent()) {
+            vertxDkimOptions.setAuid(optionsConfig.auid.get());
+        }
+
+        if (optionsConfig.bodyLimit.isPresent()) {
+            int bodyLimit = optionsConfig.bodyLimit.getAsInt();
+            if (bodyLimit <= 0) {
+                throw new ConfigurationException("Body limit must be greater than zero.");
+            }
+            vertxDkimOptions.setBodyLimit(bodyLimit);
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:84]: Just a heads-up that `io.vertx.ext.mail.DKIMSignOptions#setBodyLimit` already performs this validation :)

**Comment by ejba** [comment]
I know the sdid and the selector shouldn't be optional, but the injection test suite fails when they are strings. They will always be needed even when dkim is not enabled. I tried converting the dkim options to optional, but the properties won't load.

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java [hunk 10]
```diff
@@ -116,6 +116,17 @@ public class MailConfig {
     @ConfigItem(defaultValue = "OPTIONAL")
     public String startTLS;
 
+    /**
+     * Enable DKIM signature verification.
+     */
+    @ConfigItem(defaultValue = "false")
+    public boolean enableDkim;
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:123]: Move this flag to the `DkimSignOptionsConfig`, making the flag to enable it: `quarkus.mailer.dkim.enable=true`

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java [hunk 11]
```diff
@@ -171,6 +182,13 @@ public class MailConfig {
     @ConfigItem(defaultValue = "false")
     public boolean multiPartOnly;
 
+    /**
+     * Caches data from attachment's Stream to a temporary file.
+     * It tries to delete it after sending email.
+     */
+    @ConfigItem(defaultValue = "false")
+    public boolean cacheAttachments;
```
Comment (gastaldi) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:190]: This doesn't seem to belong to this PR?

Comment (ejba) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:190]: It affects the DKIM signing performance, see [here](https://vertx.io/docs/4.1.0/vertx-mail-client/java/#_dkim_signature_signing_emails)

Comment (gastaldi) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:190]: But can you point me to where is it being used?

Comment (ejba) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:185]: sure. https://github.com/quarkusio/quarkus/pull/25248/files#diff-f29996b22d408f02c02fac60e7c68440ad754e730f505c42836017d3b3fcdb3bR61

Comment (gastaldi) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:185]: Lol, completely missed that. Thanks. I think it needs to be produced through a build item?

Comment (ejba) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:185]: What do you mean by build item? Sorry for the question, still learning.

Comment (ejba) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:185]: Ok, got it. checking how to do it.

Comment (gastaldi) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailConfig.java:185]: I think you need to produce a `SystemPropertyBuildItem`. Here is an example: 

https://github.com/quarkusio/quarkus/blob/64173d81a8d9bb2b546481c09d480d5ff3830e0b/extensions/liquibase/deployment/src/main/java/io/quarkus/liquibase/deployment/LiquibaseProcessor.java#L85-L91

**Comment by ejba** [comment]
> I'd say go ahead and enable them, unless @cescoffier has any objections

The Windows test suite fails because it does not have the docker agent available to start the fake smtp server. So, I had to disable the integration tests.

**Comment by gastaldi** [comment]
> The Windows test suite fails because it does not have the docker agent available to start the fake smtp server. So, I had to disable the integration tests.

How about disabling the failing test only in Windows and enabling on other OSes until we figure out why the Docker agent isn't available in Windows?

**Comment by ejba** [comment]
Can you re-trigger the pipeline please?

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
Thanks!

**Comment by gastaldi** [comment]
Good job @ejba !
